File: Correlate_Bed.py
#!/usr/bin/env python

import argparse
from scipy.stats import pearsonr
from scipy.stats import spearmanr
import matplotlib.pyplot as plt
plt.switch_backend('agg') # makes sure matplotlib doesn't try to plot in the foreground


parser = argparse.ArgumentParser(description='Correlate the region values in column 5 of a bed file against another bedfile. i.e. H3K36me3 vs DNA methylation ')
parser.add_argument('--bed1', type=str,default=False,help='A normal bedfile you want to correlate regions against bed2, requires columns: chromosome\tstart\tend\tanyValue\tscore', required=True)
parser.add_argument('--bed2', type=str,help='A normal bedfile you want to correlate regions against bed1, requires columns: chromosome\tstart\tend\tanyValue\tscore', required=True)

args = parser.parse_args()
bed1=args.bed1
bed2=args.bed2

def parse_beds():
    bedlist1 = []
    bedlist2 = []
    bed1count = 0
    bed2count = 0
    matches_conditions = 0
    with open(bed1,'r') as inF1:
        lines = inF1.readlines()
        for line in lines:
            line = line.split()
            chrom1, start1, end1, score1 = str(line[0]),int(line[1]),int(line[2]),float(line[4])
            bed1count += 1
            with open(bed2,'r') as inF2:
                linez = inF2.readlines()
                for lin in linez:
                    lin = lin.split()
                    chrom2, start2, end2, score2 = str(lin[0]),int(lin[1]),int(lin[2]),float(lin[4])
                    bed2count += 1
                    #assumes a perfect match between file 1 & 2 for chromosome,position and region
                    if chrom1 == chrom2 and start1 == start2 and end1 == end2:
                        #the append is an ordered append, adding the scores inline with one another for downstream correlations
                        bedlist1.append(score1)
                        bedlist2.append(score2)
                        matches_conditions += 1 

    combined_list = [bedlist1,bedlist2]
    print("[STDOUT]: Example Bedlists:")
    print("[STDOUT]: BedList1:",bedlist1[0:5],"\n[STDOUT]: Bedlist2:",bedlist2[0:5])
    print("[STDOUT]: Out of:",bed1count,"regions in --bed1",float(100*(matches_conditions/bed1count)),"% or",matches_conditions,"exact matching regions were found in --bed2")
    print("[STDOUT]: The correlation of the measured values in these regions are as follows:")
    return combined_list

def calculate_correlation_of_matching_regions(bedlist1,bedlist2):
    # calculate spearman's correlation
    corrp, _ = pearsonr(bedlist1, bedlist2)
    print('[STDOUT]: Pearsons correlation: %.5f' % corrp)
    # calculate spearman's correlation
    corrs, _ = spearmanr(bedlist1, bedlist2)
    print('[STDOUT]: Spearmans correlation: %.5f' % corrs)
    
    return(corrp)

def plot_correlation(corrp,blist1,blist2):
    plotName = bed1[:-4] + "_vs_" + bed2[:-4] +  " Pearsons corr:" + str(round(corrp,3))
    plt.scatter(blist1,blist2)
    plt.title(plotName,)
    plt.xlabel('--bed1 values')
    plt.ylabel('--bed2 values')
    plt.grid(False)
    plot_name_jpg = bed1[:-4] + "_vs_" + bed2[:-4] + "_correlation_scatter_plot.jpg"
    plot_name_eps = bed1[:-4] + "_vs_" + bed2[:-4] + "_correlation_scatter_plot.eps"
    plt.savefig(plot_name_jpg,dpi=300)
    plt.savefig(plot_name_eps,dpi=300)
    plt.clf()


def main():
    combined_list = parse_beds()
    bedlist1, bedlist2 = combined_list[0],combined_list[1]
    corrp = calculate_correlation_of_matching_regions(bedlist1,bedlist2)
    plot_correlation(corrp,bedlist1,bedlist2)

if __name__ == "__main__":
    main()

File: CorrelateBed.jl
using ArgParse
using Statistics

function parse_commandline()
    s = ArgParseSettings(description = "Correlate the region values in column 5 of a bed file against another bedfile. i.e. H3K36me3 vs DNA methylation")
    @add_arg_table! s begin
        "--bed1"
        help = "A normal bedfile you want to correlate regions against bed2, requires columns: chromosome\tstart\tend\tanyValue\tscore"
        "--bed2"
        help = "A normal bedfile you want to correlate regions against bed1, requires columns: chromosome\tstart\tend\tanyValue\tscore"
    end
    return parse_args(s)
end

function parseBed2(bedPath)
    lines = readlines(bedPath)
    Bed2Data = Dict{String}{Float64}()
    for line in lines
        lineSplit = split(line)
        key = string(lineSplit[1]) * "_" * string(lineSplit[2]) * "_" * string(lineSplit[3])
        #println(lineSplit[5])
        Bed2Data[key] = parse(Float64,lineSplit[5])
    end
    return Bed2Data
end

function findMatches(bed1Path,bed2Data,foutName)
    bed1CorrVals = Float64[]
    bed2CorrVals = Float64[]
    count = 0
    foundMatch = 0
    for line in eachline(bed1Path)
        count += 1 
        lineSplit = split(line)
        # println(lineSplit)
        chromRegionPosition = string(lineSplit[1]) * "_" * string(lineSplit[2]) * "_" * string(lineSplit[3])
        # println(bed2Data[chromRegionPosition])
        if haskey(bed2Data,chromRegionPosition)
            foundMatch += 1
            push!(bed1CorrVals,parse(Float64,lineSplit[5]))
            push!(bed2CorrVals,bed2Data[chromRegionPosition])
        end
    end
    # provide clues as to which data was correlated and how much of it matched
    percentFound = 100*Float64(foundMatch/count)
    println("[STDOUT]: out of the total of ",count," lines in bed1, we saw perfect matches at:",foundMatch," positions or a total of: ",percentFound,"%")
    println("[STDOUT]: printing first 5 values of bed1 correlations")
    println(bed1CorrVals[1:5])
    println("[STDOUT]: printing first 5 values of bed2 correlations")
    println(bed2CorrVals[1:5])
    output = Any[bed1CorrVals,bed2CorrVals]
    return output
end

function main()
    parsed_args = parse_commandline()
    bed2Data = parseBed2(parsed_args["bed2"])
    foutName = parsed_args["bed1"][1:end-4] * ".intersect.bed"
    corVals = findMatches(parsed_args["bed1"],bed2Data,foutName)
    bed1CorVals,bed2CorVals = corVals[1],corVals[2]
    println("[STDOUT]: the correlation between bed1:",parsed_args["bed1"]," and bed2: ",parsed_args["bed2"]," is: ",Statistics.cor(bed1CorVals,bed2CorVals))
end

@time main()

#@code_native f(upperLim)
#@code_lowered f(upperLim)  # I love this, very similar to python bytecode
#@time     # also wonderful for timing pieces of code


# def parse_beds():
#     bedlist1 = []
#     bedlist2 = []
#     bed1count = 0
#     bed2count = 0
#     matches_conditions = 0
#     with open(bed1,'r') as inF1:
#         lines = inF1.readlines()
#         for line in lines:
#             line = line.split()
#             chrom1, start1, end1, TE_name = str(line[0]),int(line[1]),int(line[2]),str(line[3])
#             bed1count += 1
#             with open(bed2,'r') as inF2:
#                 linez = inF2.readlines()
#                 for lin in linez:
#                     lin = lin.split()
#                     chrom2, start2, end2, score2 = str(lin[0]),int(lin[1]),int(lin[2]),float(lin[3])
#                     bed2count += 1
#                     #assumes a perfect match between file 1 & 2 for chromosome,position and region
#                     if chrom1 == chrom2 and start1 and start2 >= start1 and end1 and end2 <= end1:
#                         #the append is an ordered append, adding the scores inline with one another for downstream correlations
#                        # bedlist1.append(score1)
#                         #bedlist2.append(score2)
#                         matches_conditions += 1
#                         print(chrom1, start1, end1, TE_name, score2, file=open("resultsfile.txt", "a"))

#     combined_list = [bedlist1,bedlist2]
#     print("[STDOUT]: Example Bedlists:")
#     print("[STDOUT]: BedList1:",bedlist1[0:5],"\n[STDOUT]: Bedlist2:",bedlist2[0:5])
#     print("[STDOUT]: Out of:",bed1count,"regions in --bed1",float(100*(matches_conditions/bed1count)),"% or",matches_conditions,"exact matching regions were found in --bed2")
#     print("[STDOUT]: The correlation of the measured values in these regions are as follows:")
#     return combined_list